synapseclient/api/entity_factory.py

```diff
@@ -0,0 +1,391 @@
+"""Factory type functions to create and retrieve entities from Synapse"""
+
+import os
+from typing import TYPE_CHECKING, Any, Dict, Optional, Union
+
+from opentelemetry import trace
+
+from synapseclient.api.entity_bundle_services_v2 import (
+    get_entity_id_bundle2,
+    get_entity_id_version_bundle2,
+)
+from synapseclient.api.entity_services import get_entities_by_md5, get_entity_path
+from synapseclient.core import utils
+from synapseclient.core.constants import concrete_types
+from synapseclient.core.download import download_file_entity_model
+from synapseclient.core.exceptions import (
+    SynapseFileNotFoundError,
+    SynapseUnmetAccessRestrictions,
+)
+
+if TYPE_CHECKING:
+    from synapseclient import Synapse
+    from synapseclient.models import File, Folder, Project
+
+
+async def get_from_entity_factory(
+    synapse_id_or_path: str,
+    version: int = None,
+    if_collision: str = "keep.both",
+    limit_search: str = None,
+    md5: str = None,
+    download_file: bool = True,
+    download_location: str = None,
+    follow_link: bool = False,
+    entity_to_update: Union["Project", "File", "Folder"] = None,
+    *,
+    synapse_client: Optional["Synapse"] = None,
+) -> Union["Project", "File", "Folder"]:
```

Comment:
Is there a plan to support Links in the future?

Reply:
That's the plan. I'll want to eventually support everything that has a synapseID here. 


We can do more to extract some functionality in this class to lower the cognitive complexity, but, we'll worry about doing that refactor when we add more supported entities here.